finetuner/tuner/pytorch/miner.py

```diff
@@ -59,6 +63,99 @@ def mine(
         diffs = matches ^ 1
 
         matches.fill_diagonal_(0)
+
+        triplets = matches.unsqueeze(2) * diffs.unsqueeze(1)
+
+        return torch.where(triplets)
+
+
+class TripletEasyHardMiner(TripletMiner):
+    def __init__(self, strategy="hard"):
+        self.strategy = strategy
+
+    def _get_per_row_min(
+        self, dist_mat: torch.Tensor, semihard_tsh: torch.Tensor = None
+    ):
+        non_zero_mask = dist_mat > 0
+
+        # Prevent wrong neg. samples from being extracted
+        row_max = torch.max(dist_mat, dim=1, keepdim=True)[0]
+        dist_mat += (row_max + 1) * torch.logical_not(non_zero_mask)
```

Comment:
Why is adding row max needed?

Reply:
Because I am getting the minimum per row. Since there are potentially zeros in the rows, I need to overwrite them with a larger value to make sure they are not extracted.